### ISSUE TITLE
Roles to form dynamic Kafka inventory groups

### ISSUE THREAD

**Issue by erikgb** [confluentinc/cp-ansible#576]
@domenicbove @JumaX We have created two roles to form dynamic inventory groups for Zookeepers (Leader/Follower) and Kafka Brokers (controller/non-controller), that we use internally i.e. to perform serial patching/rebooting of Kafka hosts. The role content is based on extracts from this project, so I wanted to discuss if this can be useful for the community. It could also potentially replace some task-lists in existing roles/playbooks.

I would be happy to submit a PR, but I thought it would best to have some initial discussions about the suggestion in context of this issue first.

Here is an initial draft, based on what we use in all our environments (currently a mix of CP 5.4 and CP 6.0): https://github.com/confluentinc/cp-ansible/compare/master...erikgb:dynamic-groups

I know it will require some rework to eventually be accepted as a PR here, but you'll hopefully get the idea....

**Comment by domenicbove** [comment]
@erikgb Very cool! Can you share a sample of a playbook that makes use of the roles?

Do you think we can insert the kafka role here: https://github.com/confluentinc/cp-ansible/blob/6.1.0-post/kafka_broker.yml
We wouldn't want duplicating code lying around.

One idea I have looking at it is to actually include the tasks file within the confluent.kafka_broker role, and call it with:
```
- import_role:
    name: confluent.kafka_broker
    tasks_from: group_hosts.yml
```

A lot of people request things like - uninstall, start, stop. and I've been thinking all those tasks files could live within the kafka_broker role, just called with their own tasks_from.

I'm glad the host ordering logic is working for you. the `group_by` module was a nice find for it.

**Comment by erikgb** [comment]
@domenicbove I am glad you seem positive! In any case I really really hope that we can make the changes with the following issue in mind: https://github.com/confluentinc/cp-ansible/issues/113. Currently we are pulling in cp-ansible using Git submodules, and it is extremely painful. Even if we you do not decide to publish to Ansible Galaxy, I hope that you somewhere down the line can make the this repository Ansible collection compatible. That would represent a **huge** improvement for us. Git submodules are just completely broken.... Currently, with Ansible 2.10, it seems impossible to import playbooks from Ansible collection, so we tend to create roles for "everything". But https://github.com/ansible/ansible/pull/67435 is now merged, so I guess this will improve from Ansible 2.11.

Of course I can share playbooks. We are (also) trying to keep things DRY, so I will need to share two playbooks. The examples are for CP 5.4. We have similar playbooks for CP 6.0, but using the new "variables" role name.

First `gather_kafka_groups.yml` which is using the roles (from an Ansible collection):
```yaml
---
- name: Group Zookeeper hosts
  hosts: zookeeper
  gather_facts: false
  tags: zookeeper
  roles:
    - role: confluent.variables_handlers
    - role: statnett.kafka_provisioning.zookeeper_groups

- name: Group Kafka Broker hosts
  hosts: kafka_broker
  gather_facts: false
  tags: kafka_broker
  roles:
    - role: confluent.variables_handlers
    - role: statnett.kafka_provisioning.kafka_broker_groups

- name: Debug cluster state
  hosts: localhost
  gather_facts: false
  tasks:
    - name: Zookeeper Leader
      debug:
        msg: "{{ groups.zookeeper_leader | default([]) | to_nice_yaml }}"

    - name: Zookeeper Followers
      debug:
        msg: "{{ groups.zookeeper_follower | default([]) | to_nice_yaml }}"

    - name: Kafka Broker Controller
      debug:
        msg: "{{ groups.kafka_broker_controller | default([]) | to_nice_yaml }}"

    - name: Kafka Broker non-Controller
      debug:
        msg: "{{ groups.kafka_broker_non_controller | default([]) | to_nice_yaml }}"

- name: Validate Zookeeper leader and Kafka Broker controller
  hosts: localhost
  tasks:
    - name: Validate leader and controller
      block:
        - name: Validate Zookeeper leader
          fail:
            msg: "[ERROR] No Zookeeper leader determined"
          when: groups.zookeeper_leader is not defined or groups.zookeeper_leader|length == 0

        - name: Validate Kafka Broker controller
          fail:
            msg: "[ERROR] No Kafka Broker controller determined"
          when: groups.kafka_broker_controller is not defined or groups.kafka_broker_controller|length == 0
      when: kafka_groups_validate_leader_controller | default(true)
```

Then `patch_system.yml`, that performs the patching of our hosts:
```yaml
---
- import_playbook: gather_kafka_groups.yml

- name: Patch system on all hosts, except Zookeeper leader and Kafka Broker controller, serially
  hosts: vm:!zookeeper_leader:!kafka_broker_controller
  gather_facts: false
  serial: 1
  roles:
    - role: confluent.variables_handlers
    - role: statnett.kafka_provisioning.patch_system

- name: Patch system on Kafka Broker controller
  hosts: kafka_broker_controller
  gather_facts: false
  serial: 1
  roles:
    - role: confluent.variables_handlers
    - role: statnett.kafka_provisioning.patch_system

- name: Patch system on Zookeeper leader
  hosts: zookeeper_leader
  gather_facts: false
  serial: 1
  roles:
    - role: confluent.variables_handlers
    - role: statnett.kafka_provisioning.patch_system
```

**Comment by domenicbove** [comment]
Ok ya looks straightforward enough to move the logic into a role. WDYT about going the "tasks_from" route?

And regarding the ansible galaxy request, I'll talk to our PM. I've been interested in some restructuring. We have a lot of good work in our backlog, so it's a matter of prioritization. If pushing to galaxy enables more usage/contributions to cp-ansible, then thats a great point to consider!

**Comment by erikgb** [comment]
@domenicbove I think the "tasks_from" route should work just fine. That will also solve some of the variable issues I have noticed in "tasks-files" and playbooks - since the "tasks_from" should also load any variables defined in context of the role. We might run into this "well known" Ansible bug (https://github.com/ansible/ansible/issues/20493) - already mentioned here: https://github.com/confluentinc/cp-ansible/pull/449#issuecomment-716622078, but I'll guess it should be possible to sort out.

Reaching Ansible Galaxy is not a must-have for us, but following the Ansible collection directory layout would be very nice indeed.

Do you want me to start on a PR to take this further? I think have now have an idea on how it might look like. Which branch should I target?

**Comment by domenicbove** [comment]
Yes since all you are proposing in this issue is a refactor I see no issue with accepting your PR. Please create it!
As far as which branch, either 6.1.x or 6.2.x. 6.2.x is for our next major release ie major changes/ features... This wouldn't be too big of a change so you could make the case for either. If you'd like it released sooner then go with 6.1.x

Thanks!

**Comment by erikgb** [comment]
Resolved by https://github.com/confluentinc/cp-ansible/pull/578
